// instagram-url-direct/dist/cjs/instagram.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var instagram_exports = {};
__export(instagram_exports, {
  instagramGetUrl: () => instagramGetUrl
});
module.exports = __toCommonJS(instagram_exports);
var import_axios = __toESM(require("axios"), 1);
var import_qs = __toESM(require("qs"), 1);
async function instagramGetUrl(url_media) {
  return new Promise(async (resolve, reject) => {
    try {
      url_media = await checkRedirect(url_media);
      const SHORTCODE = getShortcode(url_media);
      const INSTAGRAM_REQUEST = await instagramRequest(SHORTCODE);
      const OUTPUT_DATA = createOutputData(INSTAGRAM_REQUEST);
      resolve(OUTPUT_DATA);
    } catch (err) {
      let error = {
        error: err.message
      };
      reject(error);
    }
  });
}
__name(instagramGetUrl, "instagramGetUrl");
async function checkRedirect(url) {
  let split_url = url.split("/");
  if (split_url.includes("share")) {
    let res = await import_axios.default.get(url);
    return res.request.path;
  }
  return url;
}
__name(checkRedirect, "checkRedirect");
function formatPostInfo(requestData) {
  try {
    let mediaCapt = requestData.edge_media_to_caption.edges;
    const capt = mediaCapt.length === 0 ? "" : mediaCapt[0].node.text;
    return {
      owner_username: requestData.owner.username,
      owner_fullname: requestData.owner.full_name,
      is_verified: requestData.owner.is_verified,
      is_private: requestData.owner.is_private,
      likes: requestData.edge_media_preview_like.count,
      is_ad: requestData.is_ad,
      caption: capt
    };
  } catch (err) {
    throw new Error(`Failed to format post info: ${err.message}`);
  }
}
__name(formatPostInfo, "formatPostInfo");
function formatMediaDetails(mediaData) {
  try {
    if (mediaData.is_video) {
      return {
        type: "video",
        dimensions: mediaData.dimensions,
        video_view_count: mediaData.video_view_count,
        url: mediaData.video_url,
        thumbnail: mediaData.display_url
      };
    } else {
      return {
        type: "image",
        dimensions: mediaData.dimensions,
        url: mediaData.display_url
      };
    }
  } catch (err) {
    throw new Error(`Failed to format media details: ${err.message}`);
  }
}
__name(formatMediaDetails, "formatMediaDetails");
function getShortcode(url) {
  try {
    let split_url = url.split("/");
    let post_tags = ["p", "reel", "tv", "reels"];
    let index_shortcode = split_url.findIndex((item) => post_tags.includes(item)) + 1;
    let shortcode = split_url[index_shortcode];
    return shortcode;
  } catch (err) {
    throw new Error(`Failed to obtain shortcode: ${err.message}`);
  }
}
__name(getShortcode, "getShortcode");
function isSidecar(requestData) {
  try {
    return requestData["__typename"] == "XDTGraphSidecar";
  } catch (err) {
    throw new Error(`Failed sidecar verification: ${err.message}`);
  }
}
__name(isSidecar, "isSidecar");
async function instagramRequest(shortcode) {
  var _a;
  try {
    const BASE_URL = "https://www.instagram.com/graphql/query";
    const INSTAGRAM_DOCUMENT_ID = "8845758582119845";
    let dataBody = import_qs.default.stringify({
      "variables": JSON.stringify({
        "shortcode": shortcode,
        "fetch_tagged_user_count": null,
        "hoisted_comment_id": null,
        "hoisted_reply_id": null
      }),
      "doc_id": INSTAGRAM_DOCUMENT_ID
    });
    let config = {
      method: "post",
      maxBodyLength: Infinity,
      url: BASE_URL,
      headers: {
        "Content-Type": "application/x-www-form-urlencoded"
      },
      data: dataBody
    };
    const { data } = await import_axios.default.request(config);
    if (!((_a = data.data) === null || _a === void 0 ? void 0 : _a.xdt_shortcode_media))
      throw new Error("Only posts/reels supported, check if your link is valid.");
    return data.data.xdt_shortcode_media;
  } catch (err) {
    throw new Error(`Failed instagram request: ${err.message}`);
  }
}
__name(instagramRequest, "instagramRequest");
function createOutputData(requestData) {
  try {
    let url_list = [], media_details = [];
    const IS_SIDECAR = isSidecar(requestData);
    if (IS_SIDECAR) {
      requestData.edge_sidecar_to_children.edges.forEach((media) => {
        media_details.push(formatMediaDetails(media.node));
        if (media.node.is_video) {
          url_list.push(media.node.video_url);
        } else {
          url_list.push(media.node.display_url);
        }
      });
    } else {
      media_details.push(formatMediaDetails(requestData));
      if (requestData.is_video) {
        url_list.push(requestData.video_url);
      } else {
        url_list.push(requestData.display_url);
      }
    }
    return {
      results_number: url_list.length,
      url_list,
      post_info: formatPostInfo(requestData),
      media_details
    };
  } catch (err) {
    throw new Error(`Failed to create output data: ${err.message}`);
  }
}
__name(createOutputData, "createOutputData");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  instagramGetUrl
});
